File: orly/src/references.ts
import { existsSync } from "node:fs";
import { dirname, extname, isAbsolute, join, relative, resolve } from "node:path";

import { isBelow, OrlyError } from "./model";

const PACK_LINE = /^(.*?)[ \t]*<!--[ \t]*oracle-packs:([^>]+)[ \t]*-->[ \t]*$/;
const PACK_START = /^[ \t]*<!--[ \t]*oracle-packs:start ([^>]+)[ \t]*-->[ \t]*$/;
const PACK_END = /^[ \t]*<!--[ \t]*oracle-packs:end[ \t]*-->[ \t]*$/;
const CODE_FENCE = /^[ \t]*(?:```|~~~)/;
const MARKDOWN_LINK = /\[[^\]]*\]\(([^)]+)\)/g;
const DISPATCH_REFERENCE = /dispatch\/[A-Za-z0-9_.-]+\.md/g;
const COMMENT_OPEN = "<!--";
const COMMENT_CLOSE = "-->";
const NEWLINE = "\n";

export function renderProfileText(
  content: string,
  selectedPacks: Set<string>,
  knownPacks: Set<string>,
  source: string,
): string {
  return walkPackMarkers(content, selectedPacks, knownPacks, source).join(NEWLINE).trim();
}

function walkPackMarkers(
  content: string,
  selectedPacks: Set<string>,
  knownPacks: Set<string>,
  source: string,
): string[] {
  const rendered: string[] = [];
  let activeBlock: string[] | undefined;
  let includeBlock = true;
  const lines = content.split(/\r?\n/);
  for (const [index, line] of lines.entries()) {
    const lineNumber = index + 1;
    const start = line.match(PACK_START);
    if (start) {
      if (activeBlock) throw new OrlyError(`${source}:${lineNumber}: nested orly pack block`);
      activeBlock = packNames(start[1] ?? "", knownPacks, source, lineNumber);
      includeBlock = activeBlock.some((name) => selectedPacks.has(name));
      continue;
    }
    if (PACK_END.test(line)) {
      if (!activeBlock) throw new OrlyError(`${source}:${lineNumber}: unmatched orly pack block end`);
      activeBlock = undefined;
      includeBlock = true;
      continue;
    }
    if (activeBlock && !includeBlock) continue;
    const match = line.match(PACK_LINE);
    if (match) {
      const names = packNames(match[2] ?? "", knownPacks, source, lineNumber);
      if (names.some((name) => selectedPacks.has(name))) rendered.push((match[1] ?? "").trimEnd());
      continue;
    }
    rendered.push(line);
  }
  if (activeBlock) throw new OrlyError(`${source}: unclosed orly pack block`);
  return rendered;
}

// `fallbackRoot` is the repository the staged tree is about to land in. A file
// the install skipped because its source and target are the same path — the
// engine checkout updating itself — is absent from the stage but present in
// that repository, and a citation of it resolves. Without the fallback, a
// self-update fails closure against files it deliberately left alone.
export async function referenceClosureErrors(
  outputRoot: string,
  renderedPaths: string[],
  fallbackRoot?: string,
): Promise<string[]> {
  const errors = new Set<string>();
  const root = resolve(outputRoot);
  const present = (path: string): boolean =>
    existsSync(path) || (fallbackRoot !== undefined && existsSync(join(fallbackRoot, relative(root, path))));
  for (const sourcePath of renderedPaths.filter((path) => extname(path) === ".md").sort()) {
    const relativeSource = relative(outputRoot, sourcePath).replaceAll("\\", "/");
    const content = await Bun.file(sourcePath).text();
    let inComment = false;
    let inFence = false;
    for (const [index, line] of content.split(/\r?\n/).entries()) {
      const lineNumber = index + 1;
      // A fenced block is a code sample, not prose citing a file. A shell
      // one-liner carrying `sed -E 's#.*[:/]([^:/]+/[^/]+)$#\1#'` matches the
      // markdown-link shape exactly, and grading it demanded a file named
      // after the regex.
      if (CODE_FENCE.test(line)) { inFence = !inFence; continue; }
      if (inFence) continue;
      const commented = inComment;
      inComment = commentStateAfter(line, inComment);
      // A path inside an HTML comment is authoring guidance — `docs/TEMPLATE.md`
      // lists a per-surface menu of dispatch façades a spec author might cite.
      // Grading those as live citations would force every repository to carry
      // every language pack just to satisfy a comment naming the alternatives.
      if (commented || inComment) continue;
      for (const match of line.matchAll(MARKDOWN_LINK)) {
        const rawTarget = match[1] ?? "";
        const target = markdownTarget(rawTarget);
        if (!target) continue;
        const resolved = resolve(dirname(sourcePath), target);
        if (!isBelow(resolved, root)) errors.add(`snapshot reference escapes repository: ${relativeSource}:${lineNumber} -> ${rawTarget}`);
        else if (!present(resolved)) errors.add(`missing snapshot reference: ${relativeSource}:${lineNumber} -> ${rawTarget}`);
      }
      for (const target of line.matchAll(DISPATCH_REFERENCE)) {
        const path = target[0];
        if (!present(join(outputRoot, path))) errors.add(`missing dispatch reference: ${relativeSource}:${lineNumber} -> ${path}`);
      }
    }
  }
  return [...errors].sort();
}

// Whether the line leaves the scanner inside an HTML comment. Pack markers open
// and close on one line, so they never flip the state; a `<!-- tpl:` block that
// runs to a later `-->` does.
function commentStateAfter(line: string, inComment: boolean): boolean {
  let open = inComment;
  for (let index = 0; index < line.length; index += 1) {
    if (!open && line.startsWith(COMMENT_OPEN, index)) { open = true; index += COMMENT_OPEN.length - 1; continue; }
    if (open && line.startsWith(COMMENT_CLOSE, index)) { open = false; index += COMMENT_CLOSE.length - 1; }
  }
  return open;
}

function packNames(value: string, knownPacks: Set<string>, source: string, lineNumber: number): string[] {
  const names = value.split(",").map((name) => name.trim()).filter(Boolean);
  if (names.length === 0) throw new OrlyError(`${source}:${lineNumber}: orly pack marker must name a pack`);
  const unknown = names.filter((name) => !knownPacks.has(name)).sort();
  if (unknown.length > 0) throw new OrlyError(`${source}:${lineNumber}: unknown orly pack marker: ${unknown.join(", ")}`);
  return names;
}

function markdownTarget(rawTarget: string): string | undefined {
  const first = rawTarget.trim().split(/\s+/, 1)[0]?.replace(/^<|>$/g, "") ?? "";
  const target = decodeURIComponent(first.split("#", 1)[0] ?? "");
  if (!target || isAbsolute(target) || /^(https?:|mailto:|app:)/.test(target)) return undefined;
  return target;
}

File: oracle-rules/src/oracle_rules/references.py
from __future__ import annotations

import re
from pathlib import Path
from urllib.parse import unquote


PACK_LINE = re.compile(r"^(?P<body>.*?)[ \t]*<!-- oracle-packs:(?P<packs>[^>]+) -->[ \t]*$")
PACK_START = re.compile(r"^[ \t]*<!-- oracle-packs:start (?P<packs>[^>]+) -->[ \t]*$")
PACK_END = re.compile(r"^[ \t]*<!-- oracle-packs:end -->[ \t]*$")
MARKDOWN_LINK = re.compile(r"\[[^\]]*\]\(([^)]+)\)")
DISPATCH_REFERENCE = re.compile(r"dispatch/[A-Za-z0-9_.-]+\.md")


class SnapshotReferenceError(Exception):
    pass


def render_profile_text(
    content: str,
    selected_packs: set[str],
    known_packs: set[str],
    source: str,
) -> str:
    rendered: list[str] = []
    active_block: tuple[str, ...] | None = None
    include_block = True
    for line_number, line in enumerate(content.splitlines(), 1):
        start_match = PACK_START.fullmatch(line)
        if start_match:
            if active_block is not None:
                raise SnapshotReferenceError(
                    f"{source}:{line_number}: nested oracle pack block"
                )
            active_block = _pack_names(
                start_match.group("packs"), known_packs, source, line_number
            )
            include_block = bool(selected_packs.intersection(active_block))
            continue
        if PACK_END.fullmatch(line):
            if active_block is None:
                raise SnapshotReferenceError(
                    f"{source}:{line_number}: unmatched oracle pack block end"
                )
            active_block = None
            include_block = True
            continue
        if active_block is not None and not include_block:
            continue
        line_match = PACK_LINE.fullmatch(line)
        if line_match:
            line_packs = _pack_names(
                line_match.group("packs"), known_packs, source, line_number
            )
            if selected_packs.intersection(line_packs):
                rendered.append(line_match.group("body").rstrip())
            continue
        rendered.append(line)
    if active_block is not None:
        raise SnapshotReferenceError(f"{source}: unclosed oracle pack block")
    return "\n".join(rendered).strip()


def reference_closure_errors(output_root: Path, rendered_paths: list[Path]) -> list[str]:
    errors: list[str] = []
    root = output_root.resolve()
    markdown_paths = sorted(path for path in rendered_paths if path.suffix == ".md")
    for source_path in markdown_paths:
        relative_source = source_path.relative_to(output_root).as_posix()
        content = source_path.read_text(encoding="utf-8")
        for line_number, line in enumerate(content.splitlines(), 1):
            for raw_target in MARKDOWN_LINK.findall(line):
                target = _markdown_target(raw_target)
                if target is None:
                    continue
                resolved = (source_path.parent / target).resolve()
                if not _is_below(resolved, root):
                    errors.append(
                        f"snapshot reference escapes repository: "
                        f"{relative_source}:{line_number} -> {raw_target}"
                    )
                elif not resolved.exists():
                    errors.append(
                        f"missing snapshot reference: "
                        f"{relative_source}:{line_number} -> {raw_target}"
                    )
            if relative_source == "AGENTS.md":
                for target in DISPATCH_REFERENCE.findall(line):
                    if not (output_root / target).is_file():
                        errors.append(
                            f"missing dispatch reference: "
                            f"{relative_source}:{line_number} -> {target}"
                        )
    return sorted(set(errors))


def _pack_names(
    value: str, known_packs: set[str], source: str, line_number: int
) -> tuple[str, ...]:
    names = tuple(name.strip() for name in value.split(",") if name.strip())
    if not names:
        raise SnapshotReferenceError(
            f"{source}:{line_number}: oracle pack marker must name a pack"
        )
    unknown = sorted(set(names) - known_packs)
    if unknown:
        raise SnapshotReferenceError(
            f"{source}:{line_number}: unknown oracle pack marker: {', '.join(unknown)}"
        )
    return names


def _markdown_target(raw_target: str) -> str | None:
    target = raw_target.strip().split(maxsplit=1)[0].strip("<>")
    target = unquote(target.split("#", 1)[0])
    if not target or target.startswith(("/", "http://", "https://", "mailto:", "app://")):
        return None
    return target


def _is_below(path: Path, root: Path) -> bool:
    try:
        path.relative_to(root)
    except ValueError:
        return False
    return True
